### PR TITLE
Automatic plugin report generator

### PR DESCRIPTION
**Pull request by ericwill**
Contributes a script that:
- Iterates through all entries in the `vscode-extensions.json` file
- Checks if any plugins listed there are out of date
- Generates a table in a markdown file, listing all plugins and highlighting which ones need an update

This script is meant to be run daily in a GitHub action job -- PR to follow.

Signed-off-by: Eric Williams <ericwill@redhat.com>

### CODE REVIEW

File: tools/automation/check-plugin-updates.js [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+function readPackageJSON(pathToPackageJSON) {
+  let rawData = fs.readFileSync(pathToPackageJSON, 'utf-8');
+  return JSON.parse(rawData);
+}
+
+/**
+ * @return {String} Date and time in 'YYYY-MM-DD HH-MM' format.
+ */
+function getTimeDateToday() {
+  var d = new Date(),
```
Comment (benoitf) [inline tools/automation/check-plugin-updates.js:30]: FYI you could use moment library

**Comment by ericwill** [comment]
https://github.com/eclipse/che-plugin-registry/pull/499 needs to be merged as a pre-requisite for this PR.

File: tools/automation/package.json [hunk 2]
```diff
@@ -0,0 +1,23 @@
+{
+  "name": "plugin-build-update-automation",
+  "version": "0.0.0",
+  "description": "Scripts for updating, adding, removing, and building plugins in the Che Plugin Registry",
+  "main": "check-plugin-updates.js",
+  "scripts": {
```
[the remark this answers is not in the document]
Comment (ericwill) [reply tools/automation/package.json:6]: No I suppose not, I just adapted the update script OpenVSX is using. I can use TypeScript instead.

File: tools/automation/check-plugin-updates.js [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+  var time = d.getUTCHours() + ":" + (minutes < 10 ? "0" + minutes : minutes);
+  return date + " " + time;
+}
+
+/**
+ * Removes the '/tmp/repository' folder
+ */
+function cleanUpTempRepo() {
+  try {
+    execSync('rm -rf /tmp/repository', {stdio : 'pipe'});
```
[the remark this answers is not in the document]
Comment (ericwill) [reply tools/automation/check-plugin-updates.js:49]: My JavaScript isn't super strong, but isn't it the same end result? Or is calling sync functions something one isn't supposed to do?

[the remark this answers is not in the document]
Comment (benoitf) [reply tools/automation/check-plugin-updates.js:23]: In this case it's to not rely on bash/unix tools

File: tools/automation/check-plugin-updates.js [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+const fs = require('fs');
+const path = require('path');
+const semver = require('semver');
+const execSync = require('child_process').execSync;
+
+/**
+ * @param {string | number | Buffer | import("url").URL} pathToPackageJSON
+ * @return {object} Parsed package.json file.
+ */
+function readPackageJSON(pathToPackageJSON) {
```
[the remark this answers is not in the document]
Comment (ericwill) [reply tools/automation/check-plugin-updates.js]: This seems to cause an issue where it caches the first `package.json`

File: tools/automation/check-plugin-updates.js [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+    continue;
+  }
+
+  // Clean up repository folder
+  cleanUpTempRepo();
+}
+
+// Write the file
+try {
+  fs.writeFileSync('./../../index.md', report);
```
[the remark this answers is not in the document]
Comment (ericwill) [reply tools/automation/check-plugin-updates.js:88]: [writeFile is utf-8 by default](https://nodejs.org/api/fs.html#fs_fs_writefilesync_file_data_options), and is there any reason to make it async? There is no more code after it.

[the remark this answers is not in the document]
Comment (benoitf) [reply tools/automation/check-plugin-updates.js:95]: well, it's just that the sync methods were designed at a time where nobody used Promises :)

just it's cleaner/more parallel to not hold the event loop especially with async/await keywords

File: tools/automation/package.json (hunk 2 above)
[the remark this answers is not in the document]
Comment (benoitf) [reply tools/automation/package.json:6]: yep, I think we could benefit from typesafing checks.

**Comment by benoitf** [comment]
Hi, may I help to merge it faster ?

**Comment by ericwill** [comment]
> Hi, may I help to merge it faster ?

Sure, if you like. :)

**Comment by benoitf** [comment]
what are the missing steps where I can help ?

**Comment by ericwill** [comment]
> what are the missing steps where I can help ?

The only thing I would appreciate some pointers on is redoing it in TypeScript, everything else was just waiting on the `vscode-extensions.json` file being merged and having the schema finalized. Now that that's done I can proceed.

**Comment by ericwill** [comment]
I've adapted the current script to make it async + use the new `vscode-extensions.json` schema. I spent sometime trying a TypeScript approach and it didn't go very far, plus I think it's a bit overkill for a script like this.

@benoitf If you consider it a blocker I can do it in TypeScript, I'll just need a bit more time due to PTO/sprint planning coming up. We can also merge the current approach and leave the TypeScript conversion as a TODO.

**Comment by benoitf** [comment]
@ericwill https://github.com/ericwill/che-plugin-registry/pull/1

**Comment by ericwill** [comment]
Thanks for Florent for the contribution. Let me update the README and then I think we can merge this one.

**Pull request merged**